man/adoc/bpftrace.adoc

```diff
@@ -977,6 +977,131 @@ The following snippets create a map with key signature `(int64, string[16])` and
 @[(pid, comm)]++
 ----
 
+==== Hygienic macros
+
+Hygienic macros (usually just "macros") provide a way for you to structure your script. They can be
+useful when you want to factor out code into smaller, more understandable parts. Or if you want to
+share code between probes.
+
+At a high level, macros can be thought of as semantic aware text replacement. They accept (optional)
+variable and map arguments - the body of the macro may only access maps and variables passed in through
+the arguments (the hygiene) or literal expressions. The body of the macro is exactly one block expression.
+
+For example, these are valid usages of macros:
+
+----
+config = {
+    unstable_macro=1;
+}
+
+macro one() {
+  1
+}
+
+macro add_one($other_name) {
+  $other_name + 1
+}
+
+macro add_one_to_each($a, @b) {
+  $a + 1 + @b + 1
+}
+
+macro side_effect($x) {
+  print($x)
+}
+
+macro add_two($x) {
+  add_one($x) + 1
+}
+
+BEGIN {
+  print(one());
+
+  $a = 42;
+  print(add_one($a));
+
+  @b = 3;
+  print(add_one_to_each($a, @b));
+
+  side_effect(5)
+
+  print(add_two(1));
+}
+
+END {
+  clear(@b);
+}
+
+/*
+ * Sample output:
+ * Attaching 2 probes...
+ * 1
+ * 43
+ * 47
+ * 5
+ * 3
+ */
+----
+
+Some examples of invalid macro usage:
+
+----
+config = {
+    unstable_macro=1;
+}
+
+macro not_expression() {
+  $var = 1;                    // BAD: Not an expression
+}
+
+macro unhygienic_access() {
+  @x++;                        // BAD: @x not passed in
+}
+
+macro wrong_parameter_type($x) {
+  $x++;
+}
+
+BEGIN {
+  @x = 1;
+  unhygienic_access();
+
+  wrong_parameter_type(@x);    // BAD: macro expects a scratch variable
+}
+----
+
+Macros also allow variable declarations but not if they shadow variables
```

Comment:
Ohhhh, this one is tricky. Do we need to automatically rewrite these variables? If you are authoring a macro, you can't really be subject to constraints about what's defined in the *caller* context, e.g. this would make defining macros for the standard library nearly impossible.

Reply:
Yeah, re-writing is a good idea and I'm already doing something like that when passing literals. I could also just re-write variable assignments if we detect that the variable isn't included in the macro params, which would prevent un-hygienic errors. Let me look into doing that.

Reply:
If I understand the code correctly, the re-writing is already implemented so this statement no longer holds, does it?